Disable weak encryption using installer commands (#311)
theforeman/foreman-documentation: guides/doc-Administering_Red_Hat_Satellite/topics/Disabling_Weak_Encryption.adoc

[[Disabling_Weak_SSL_2-0_and_3-0_Encryption]]
== Disabling TLS 1.0 and TLS 1.1 Encryption

You might want to change the encryption settings for {Project} depending on the security requirements of your infrastructure or to fix vulnerabilities quickly.

Apache and Qpid services in {Project} have TLS 1.0 and 1.1 encryption enabled by default. Use this procedure on {Project} and {SmartProxy} to configure {Project} and {SmartProxy} to only allow TLS 1.2, as required, for example, by PCI-DSS regulations.

.Procedure

. Open the `/etc/foreman-installer/custom-hiera.yaml` file for editing:
+
[options="nowrap" subs="+quotes,attributes"]
----
# vi /etc/foreman-installer/custom-hiera.yaml
----

> . Add the following entry:

Reviewer: It's 2 entries again
```suggestion
. Add the following entries:
```

+
[options="nowrap"]
----
# Apache
apache::mod::ssl::ssl_protocol: [ 'ALL' , '-SSLv3' , '-TLSv1' , '-TLSv1.1' , '+TLSv1.2' ]

# QPID Dispatch
foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'
----

. Enter the `{foreman-installer}` command to configure Qpid to use TLS 1.2 only:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {foreman-installer} \

> --foreman-proxy-content-qpid-router-ssl-protocols=TLSv1.2

Reviewer: So I noticed that when the ciphers were tightened it also stopped listening on TLS 1.0 and 1.1 which would make that redundant. Should we drop this line?

Author: Do you say that we do not need to run the installer at all and that the config lines in custom-hiera.yaml would do the trick? If so, I can remove this step completely and return the step that restarts the services.

Reviewer: No, still run the installer but without the argument. Running the installer is crucial to actually making the config file changes.

Author: Did you notice that I removed the `foreman_proxy_content::qpid_router_ssl_protocols: [ 'TLSv1.2' ]` config line? Now we say to only set the `foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'` configuration. Will that do the trick? Want to check before committing it.

Reviewer: It appeared to be when I tried it out. Now I'm not sure if it was just nmap's reporting. Will try to figure it out.

If you're unsure, you can do both for now and we can simplify it in a future iteration.

----

> . Restart the `{foreman-maintain}` services:
> +
> [options="nowrap" subs="+quotes,attributes"]
> ----
> # {foreman-maintain} service restart
> ----

Reviewer: The installer should restart any services. Not sure why you would restart them again after.
```suggestion
```

[[Disabling_Weak_SSL_2-0_and_3-0_Encryption]]

> === Disabling Weak SSL 2.0 and SSL 3.0 Encryption

Reviewer: ```suggestion
=== Disabling TLS 1.0 and 1.1 Encryption
```

Author: Done.

> Apache and Qpid services in {Project} have SSL encryption disabled by default, however, TLS 1.0 and 1.1 encryption is enabled. If your {Project} and {SmartProxy} fail Nessus scans because of encryption vulnerabilities, or your security infrastructure requires that you disable TSL 1.0 and TSL 1.1, complete this procedure on {Project} and {SmartProxy} to remove weak encryption.

Reviewer: `SSL encryption` is ambiguous. In many places SSL and TLS are used interchangably. I would not mention SSLv2 and SSLv3 at all. In 2020 no product should even think about shipping them IMHO.
```suggestion
Apache and Qpid services in {Project} have TLS 1.0 and 1.1 encryption enabled by default. This procedure tightens the security to only allow TLS 1.2, as required in various places (such as PCI-DSS regulations).
```

Author: Done.

[[Disabling_Weak_SSL_2-0_and_3-0_Encryption]]
=== Disabling Weak SSL 2.0 and SSL 3.0 Encryption

> Apache and Qpid services in {Project} use SSL 2.0 and SSL 3.0 encryption by default. If your {Project} or {SmartProxy} fails Nessus scans because of SSL vulnerabilities, or your security infrastructure requires that you disable SSL 2.0 and SSL 3.0, complete this procedure on {Project} or {SmartProxy} to remove weak encryption.

Author: @ekohl can you help to verify the first sentence here. Is that correct that Apache and Qpid use SSL 2.0 and 3.0?
You suggested changing the title, can you elaborate on how to change it exactly?

Author: I have a default installation of Satellite 6.7. I see the following on it: 
```
PORT    STATE SERVICE
443/tcp open  https
| ssl-enum-ciphers: 
|   SSLv3: No supported ciphers found
|   TLSv1.0: 
|     ciphers: 
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_DHE_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|     compressors: 
|       NULL
|   TLSv1.1: 
|     ciphers: 
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_DHE_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|     compressors: 
|       NULL
|   TLSv1.2: 
|     ciphers: 
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 - strong
|       TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 - strong
|       TLS_DHE_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 - strong
|       TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 - strong
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 - strong
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 - strong
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 - strong
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA - strong
|       TLS_RSA_WITH_AES_128_CBC_SHA256 - strong
|       TLS_RSA_WITH_AES_128_GCM_SHA256 - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA - strong
|       TLS_RSA_WITH_AES_256_CBC_SHA256 - strong
|       TLS_RSA_WITH_AES_256_GCM_SHA384 - strong
|     compressors: 
|       NULL
|_  least strength: strong
```
I see that qpid is running when entering `systemctl status qpidd`. Sounds like it does not use SSL by default as well.

Reviewer: It would have surprised me if it did, especially SSL v2 which is really really ancient. Even SSLv3 was considered insecure back in 2014 which is already 6 years ago. If RHEL shipped something insecure by default, that would have surprised me.

Author: Shall we then dump this whole chapter 7. Disabling Weak Encryption because it is irrelevant is the current RHEL?

Author: Wait, 7.2. Disabling 64-bit Block Size Cipher Suites (SWEET32) is still relevant.

Reviewer: > Shall we then dump this whole chapter 7. Disabling Weak Encryption because it is irrelevant is the current RHEL?

I'd say that currently the instructions are valid for some. Various institutions only allow TLS 1.2 (I think PCI requires it?). Since the default install allows TLS 1.0, that is still useful. However, it should be clearly stated that its disabling TLS 1.0 and 1.1, not SSL v2/v3.

Reviewer: > Wait, 7.2. Disabling 64-bit Block Size Cipher Suites (SWEET32) is still relevant.

I think your example proves it doesn't. I don't see DES ciphers in there. Looks like it can be dropped.

Author: Are you sure that it can be dropped? Ganesh has a customer who used this in Satellite 6.5, see [BZ#1683832](https://bugzilla.redhat.com/show_bug.cgi?id=1683832).

Reviewer: I can't speak for 6.5 since that's rather ancient. It should have the changes. Note that the BZ is about making the changes persistent. However, I think that the entire instructions are no longer needed.

Reviewer: Used an internal reference Satellite 6.5 and that is not vulnerable to it. It was relevant on older releases but those are now EOL.

Author: I am more than happy to dump it then

+
[options="nowrap"]
----
# Foreman Proxy
foreman_proxy::tls_disabled_versions: [ '1.1' ]

# Dynflow

> foreman_proxy::plugin::dynflow::tls_disabled_versions: [ '1.1' ]

Reviewer: Redundant since https://projects.theforeman.org/issues/29624 which dropped TLS 1.1

> apache::mod::ssl::ssl_protocol: [ 'ALL' , '-SSLv3' , '-TLSv1' , '-TLSv1.1' , '+TLSv1.2' ]

Author: Following the Chesterton’s Fence principle, I would leave this here for now.

# Apache
apache::mod::ssl::ssl_protocol: [ 'ALL' , '-SSLv3' , '-TLSv1' , '-TLSv1.1' , '+TLSv1.2' ]

# Tomcat / Candlepin

> candlepin::tls_versions: [ '1.2' ]

Reviewer: FYI: since 2.1 this is redundant: https://projects.theforeman.org/issues/29625

# QPID Dispatch
foreman_proxy_content::qpid_router_ssl_protocols: [ 'TLSv1.2' ]

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Author: Does the `--foreman-proxy-content-qpid-router-ssl-protocols=TLSv1.2` option replaces this config line?

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: No, only the line above.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: Turns out that the default installation needs this. I think that this line should even be the default. Without it, qpid is vulnerable to SWEET32. Now I am wondering if this is sufficient. We have a qpid router, but there's also another qpid service listening. That may need a similar line.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Author: I will not remove these lines and return them then

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: I think it's better to use the command line option, similar to the qpid router ssl protocols.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Author: Now this PR adds the following command:
```
# {foreman-installer} \
--foreman-proxy-content-qpid-router-ssl-protocols=TLSv1.2
```
The Bug this PR is opened for also suggests to add the option `--foreman-proxy-tls-disabled-versions=1.1` but you said it is not required, so I removed it. The qpid-router option is still there.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: The bug is correct that the command line option is `--foreman-proxy-tls-disabled-versions`. However, the need to pass that option was removed.

> I think it's better to use the command line option, similar to the qpid router ssl protocols.

What I meant was that you can pass `--foreman-proxy-content-qpid-router-ssl-ciphers '...'` which is better than setting it in `custom-hiera.yaml`.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Author: I do not have such an option on Satellite 6.7, I will find some Satellite 6.8 instance and check there next week.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: It should be there. did you check `--full-help`?

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Author: My bad, the option is there under full-help. I cannot figure out how to use it, I tried the original syntax from the custom-hiera.yml and separating by commas, neither worked:
```
# satellite-installer --foreman-proxy-content-qpid-router-ssl-ciphers 'ALL, !aNULL, +HIGH, -SSLv3, !IDEA-CBC-SHA'
ERROR: Scenario (/etc/foreman-installer/scenarios.d/SLv3,.yaml) was not found, can not continue
# satellite-installer --foreman-proxy-content-qpid-router-ssl-ciphers 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'
ERROR: Scenario (/etc/foreman-installer/scenarios.d/SLv3:!IDEA-CBC-SHA.yaml) was not found, can not continue
```

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: I think you may be hitting [Bug #30218](https://projects.theforeman.org/issues/30218) where `-S` (in `-SSLv3`) is interpreted as a scenario switch. The linked BZ suggests it hasn't landed in downstream yet.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Reviewer: Created https://github.com/theforeman/puppet-foreman_proxy_content/pull/301 to set it out of the box. Probably doesn't help you for now though. Perhaps we should keep this in `custom-hiera.yaml` for now until we have it solved.

> foreman_proxy_content::qpid_router_ssl_ciphers: 'ALL:!aNULL:+HIGH:-SSLv3:!IDEA-CBC-SHA'

Author: Thank you, @ekohl. I have removed the config line that configures qpid tls version because this is done via the command line option, I have kept the config line that configures SSL cyphers.

> # {foreman-installer} --foreman-proxy-tls-disabled-versions=1.1 \

Reviewer: TLS 1.1 is gone since Foreman Proxy 2.1 (https://github.com/theforeman/smart-proxy/commit/ebdde78c329d6bf67b198446dae64d8952d0ac4a). Please drop this.

Author: The option `foreman-proxy-tls-disabled-versions=1.1` disables TLS version 1.1. From the help output:
`--foreman-proxy-tls-disabled-versions  List of TLS versions that will be disabled from the default`

Reviewer: But it's redundant. If you don't specify it, the Smart Proxy won't use it either. The only use it has now is to disable TLS 1.2 and only use TLS 1.3, but that's not really relevant for most deployment.

Author: Got it, thank you for the explanation. What with Satellites that exist longer, say for those who upgrade from 6.4. Will TLS 1.1 be enabled on them? Does installer disable TLS 1.1 when upgrading?

Reviewer: It will still configure it but the smart proxy application **always** disables some protocols:
https://github.com/theforeman/smart-proxy/blob/7c818d0cf1606a7d16c8334ab4fd256370eac365/lib/launcher.rb#L73-L79
Currently that's SSLv2, SSLv3, TLS 1.0 and (since Foreman 2.1 / Satellite 6.8) TLS 1.1. It then parses the disabled option:
https://github.com/theforeman/smart-proxy/blob/7c818d0cf1606a7d16c8334ab4fd256370eac365/lib/launcher.rb#L81-L90
So effectively what happens is:
```ruby
options |= OpenSSL::SSL::OP_NO_TLSv1_1 # hardcoded
options |= OpenSSL::SSL::OP_NO_TLSv1_1 # via options
```

Since it doesn't do anything, I didn't bother cleaning up options in a migration. The user will still only use strong encryption.

Author: Thank you for the clarification, I have removed this option.
